cloudinit/stages.py

```diff
@@ -397,6 +397,12 @@ def _store_rawdata(self, data, datasource):
             data = b''
         util.write_file(self._get_ipath('%s_raw' % datasource), data, 0o600)
 
+    def _store_raw_vendordata(self, data, datasource):
+        # This data may be a list or dict, convert it to a string if so
+        if type(data) in [list, dict]:
+            data = util.json_dumps(data)
```

Comment:
a few things here:
a.) I think i'd rather fail on a dict (with a good error message). As far as I can see, no datasources should be storing a dict in vendordata_raw.  Since it was really *supposed* to be bytes, lets not slip any further into letting more types in.
b.) You probably have to allow for None.
c.) you should error on any other type. If you happened to have an int there, then it will go on through to store_rawdata, which will probably error with a less useful message.

So, how about:

```python
def _store_raw_vendordata(self, data, datasource):
    # vendordata_raw should only be one of None, list, bytes, or string.
    if data is None:
        data = b''
    elif isinstance(data, list):
        data = util.json_dumps(data)
    elif isinstance(data, bytes):
        pass
    elif isinstance(data, string):
        data = data.encode('utf-8')
    else:
        raise TypeError("vendordata_raw is unsupported type '%s'", type(data))
    self._store_rawdata(data, datasource)
```

Reply:
```
def _store_raw_vendordata(self, data, datasource):
        # Only these data types
        if type(data) not in [bytes, string, list, None]:
            raise TypeError("vendordata_raw is unsupported type '%s'", type(data))
        # This data may be a list, convert it to a string if so
        if isinstance(data, list):
            data = util.json_dumps(data)
        self._store_rawdata(data, datasource)
```

This should do the same thing. Don't really need,

```
    if data is None:
        data = b''
```

This is done in _store_rawdata already.

```
elif isinstance(data, string):
        data = data.encode('utf-8')
```

Not sure about this line either, considering that,

util.py
```
content = encode_text(content)
```

was throwing the error in the first place in write_files, and thats just calling .encode with the benefit of being able to specify and, encoding='utf-8', so its the default as well.

So let me know if you still want those lines in there as it looks like it my just be duplicating code.

Reply:
Yours looks good now.   thank you.